rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+
+But the size of MemTable is limited and might be flushed & rotated at any time, it can not buffer ALL the changes since the earliest running transaction. To deal with this limitation, RocksDB's solution is also very straghtforward:
+
+1. Track the earliest sequence number of keys in the MemTable.
+2. When transaction commits, if the MemTable's earliest sequence number is bigger than the transaction's sequence number, then the transaction is considered as `Expired`.
+
+The good part of this approach is that it's very efficient and simple, and it's free of floating garbage. The bad part is that it may abort the transaction when it's not necessary, and prones to encourage the users to set a bigger MemTable size.
+
+Given the wide adaption of RocksDB & the simplicity of the implementation, we could consider to implement the conflict checking in the similar way in SlateDB.
+
+## Conflict Checking: Serializable Snapshot Isolation
+
+The difference between SSI and SI is simple: not only check the conflicts between the current transactions's Writes with others' Writes, but also check the conflicts between the current transaction's Reads with others' Writes.
+
+The key idea is to track the recent committed transactions globally, and let the current transaction checks conflicts with them, belike:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+type KeyFingerPrint = u32;
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<KingFingerPrint>,
+    read_keys: HashSet<KeyFingerPrint>,
+    committed_seq: Option<u64>
+}
+```
+
+`Oracle` is considered as a global singleton in the DB, and it tracks the recent committed transactions in the `recent_committed_txns` deque. When a transaction commits, it should push itself into the `recent_committed_txns` deque, and the entries in `recent_committed_txns` can be GCed after they are not needed, we'll cover the details on GC later.
+
+Another detail worth to be mentioned is that we do not have to track the full keys in `String` on `read_keys` and `write_keys`, which might be not memory efficient. Instead, we could just store the integer hash `KeyFingerPrint` of each key. There might introduce some unnecessary conflicts when different keys happens to got the same hash value, but the probability is extreme low.
+
+Besides the started sequence number of the transaction, we also need to track the sequence number of the transaction committed. During the execution of a transaction, some other transactions might commit, and the current transaction should check the conflicts with these committed transactions.
+
+The conflict check simply checks whether the Read & Write keys of the current transaction are overlapped with the Write keys from related recent committed transactions.
```

Comment:
Can a section on testing be added? Testing would be one of the harder part of a feature like this.